### ISSUE TITLE
Bug: Wave does not support 'null' container engine

### ISSUE THREAD

**Issue by adamrtalbot** [nextflow-io/nextflow#3965]
## Bug report 

When running nf-core/sarek on AWS via Tower (with Wave and Fusion enabled), the error `Wave does not support 'null' container engine` occurred while executing process. 

### Expected behavior and actual behavior

Command launched by Tower:

```
nextflow run 'https://github.com/nf-core/sarek'
		 -name lethal_shannon
		 -params-file 'https://api.tower.nf/ephemeral/uivdKNFGgFNqOBSBZroySA.json'
		 -with-tower
		 -r 3.1.2
		 -profile test
```

### Steps to reproduce the problem

After retrying the pipeline, it worked fine so looks like an intermittent fault.

### Program output 

[nf-4WLnF5thJHWzde.txt](https://github.com/nextflow-io/nextflow/files/11542577/nf-4WLnF5thJHWzde.txt)
[nf-4WLnF5thJHWzde.log](https://github.com/nextflow-io/nextflow/files/11542578/nf-4WLnF5thJHWzde.log)

Relevant lines:
```
May-23 02:40:29.285 [main] DEBUG nextflow.processor.TaskProcessor - Starting process > NFCORE_SAREK:SAREK:MULTIQC
May-23 02:40:29.285 [main] DEBUG nextflow.script.ScriptRunner - > Awaiting termination 
May-23 02:40:29.285 [main] DEBUG nextflow.Session - Session await
May-23 02:40:29.477 [Actor Thread 17] DEBUG nextflow.sort.BigSort - Sort completed -- entries: 1; slices: 1; internal sort time: 0.002 s; external sort time: 0.014 s; total time: 0.016 s
May-23 02:40:29.477 [Actor Thread 30] DEBUG nextflow.sort.BigSort - Sort completed -- entries: 1; slices: 1; internal sort time: 0.002 s; external sort time: 0.014 s; total time: 0.016 s
May-23 02:40:29.691 [Actor Thread 17] DEBUG nextflow.file.FileCollector - Saved collect-files list to: /seqeralabs-showcase/work/sarek/work-20230523/collect-file/4061ae76e9a565e648322056d3a4b0f1
May-23 02:40:29.704 [Actor Thread 17] DEBUG nextflow.file.FileCollector - Deleting file collector temp dir: /tmp/nxf-3322981256689798888
May-23 02:40:30.181 [Actor Thread 30] DEBUG nextflow.file.FileCollector - Saved collect-files list to: /seqeralabs-showcase/work/sarek/work-20230523/collect-file/15cfad1d2b2335369f0c21c4bc5cae4e
May-23 02:40:30.184 [Actor Thread 30] DEBUG nextflow.file.FileCollector - Deleting file collector temp dir: /tmp/nxf-3817163844287101138
May-23 02:40:30.240 [FileTransfer-1] DEBUG nextflow.file.FilePorter - Copying foreign file https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/genome/genome.fasta to work dir: s3://seqeralabs-showcase/work/sarek/work-20230523/stage-641f5c2a-6da7-4cfb-bc9a-0ffe1e9bdd7c/34/ce40d8c726425495ec5652abf4fe1f/genome.fasta
May-23 02:40:30.242 [FileTransfer-2] DEBUG nextflow.file.FilePorter - Copying foreign file https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/genome/vcf/gnomAD.r2.1.1.vcf.gz to work dir: s3://seqeralabs-showcase/work/sarek/work-20230523/stage-641f5c2a-6da7-4cfb-bc9a-0ffe1e9bdd7c/8a/317b8d38cbe2c37f63363b247d6ac4/gnomAD.r2.1.1.vcf.gz
May-23 02:40:30.242 [FileTransfer-3] DEBUG nextflow.file.FilePorter - Copying foreign file https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/genome/genome.interval_list to work dir: s3://seqeralabs-showcase/work/sarek/work-20230523/stage-641f5c2a-6da7-4cfb-bc9a-0ffe1e9bdd7c/ed/77ee7aa01066d4f21935fda7fa2c83/genome.interval_list
May-23 02:40:30.316 [FileTransfer-4] DEBUG nextflow.file.FilePorter - Copying foreign file https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/genome/vcf/mills_and_1000G.indels.vcf.gz to work dir: s3://seqeralabs-showcase/work/sarek/work-20230523/stage-641f5c2a-6da7-4cfb-bc9a-0ffe1e9bdd7c/0d/0c6b72fc09abc64bcedc57c3945f5b/mills_and_1000G.indels.vcf.gz
May-23 02:40:30.419 [FileTransfer-5] DEBUG nextflow.file.FilePorter - Copying foreign file https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/genome/vcf/dbsnp_146.hg38.vcf.gz to work dir: s3://seqeralabs-showcase/work/sarek/work-20230523/stage-641f5c2a-6da7-4cfb-bc9a-0ffe1e9bdd7c/72/e48832bf2403a976af26b80074ac39/dbsnp_146.hg38.vcf.gz
May-23 02:40:30.423 [FileTransfer-6] DEBUG nextflow.file.FilePorter - Copying foreign file https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/fastq/test_1.fastq.gz to work dir: s3://seqeralabs-showcase/work/sarek/work-20230523/stage-641f5c2a-6da7-4cfb-bc9a-0ffe1e9bdd7c/ee/df537335f333a1ef4e969a7d8456da/test_1.fastq.gz
May-23 02:40:30.451 [Actor Thread 16] DEBUG i.s.wave.plugin.config.WaveConfig - Wave strategy not specified - using default: [container, dockerfile, conda]
May-23 02:40:30.453 [FileTransfer-7] DEBUG nextflow.file.FilePorter - Copying foreign file https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/fastq/test_2.fastq.gz to work dir: s3://seqeralabs-showcase/work/sarek/work-20230523/stage-641f5c2a-6da7-4cfb-bc9a-0ffe1e9bdd7c/2f/f7039aa73e1383d79347c81ed4d378/test_2.fastq.gz
May-23 02:40:30.456 [Actor Thread 3] DEBUG nextflow.util.CacheHelper - [WARN] Unknown hashing type: class nextflow.util.MemoryUnit
May-23 02:40:30.459 [Actor Thread 16] DEBUG io.seqera.wave.plugin.WaveClient - Wave server endpoint: https://wave.seqera.io
May-23 02:40:30.470 [Actor Thread 13] DEBUG nextflow.processor.TaskProcessor - Handling unexpected condition for
  task: name=NFCORE_SAREK:SAREK:PREPARE_GENOME:SAMTOOLS_FAIDX; work-dir=null
  error [java.lang.IllegalArgumentException]: Wave does not support 'null' container engine
May-23 02:40:30.470 [Actor Thread 50] DEBUG nextflow.processor.TaskProcessor - Handling unexpected condition for
  task: name=NFCORE_SAREK:SAREK:PREPARE_GENOME:GATK4_CREATESEQUENCEDICTIONARY; work-dir=null
  error [java.lang.IllegalArgumentException]: Wave does not support 'null' container engine
May-23 02:40:30.471 [Actor Thread 38] DEBUG nextflow.processor.TaskProcessor - Handling unexpected condition for
  task: name=NFCORE_SAREK:SAREK:PREPARE_GENOME:BWAMEM1_INDEX; work-dir=null
  error [java.lang.IllegalArgumentException]: Wave does not support 'null' container engine
May-23 02:40:30.474 [Actor Thread 18] DEBUG nextflow.processor.TaskProcessor - Handling unexpected condition for
  task: name=NFCORE_SAREK:SAREK:PREPARE_GENOME:GATK4_CREATESEQUENCEDICTIONARY; work-dir=null
  error [java.lang.InterruptedException]: java.lang.InterruptedException
May-23 02:40:30.474 [Actor Thread 16] DEBUG nextflow.processor.TaskProcessor - Handling unexpected condition for
  task: name=NFCORE_SAREK:SAREK:PREPARE_GENOME:TABIX_KNOWN_INDELS (1); work-dir=null
  error [java.lang.IllegalArgumentException]: Wave does not support 'null' container engine
May-23 02:40:30.475 [Actor Thread 7] DEBUG nextflow.processor.TaskProcessor - Handling unexpected condition for
  task: name=NFCORE_SAREK:SAREK:PREPARE_GENOME:TABIX_KNOWN_INDELS; work-dir=null
  error [java.lang.InterruptedException]: java.lang.InterruptedException
May-23 02:40:30.480 [Actor Thread 15] DEBUG nextflow.processor.TaskProcessor - Handling unexpected condition for
  task: name=NFCORE_SAREK:SAREK:PREPARE_GENOME:SAMTOOLS_FAIDX; work-dir=null
  error [java.lang.InterruptedException]: java.lang.InterruptedException
May-23 02:40:30.483 [Actor Thread 38] ERROR nextflow.processor.TaskProcessor - Error executing process > 'NFCORE_SAREK:SAREK:PREPARE_GENOME:BWAMEM1_INDEX'

Caused by:
  Wave does not support 'null' container engine

java.lang.IllegalArgumentException: Wave does not support 'null' container engine
	at io.seqera.wave.plugin.resolver.WaveContainerResolver.resolveImage(WaveContainerResolver.groovy:88)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.codehaus.groovy.runtime.callsite.PlainObjectMetaMethodSite.doInvoke(PlainObjectMetaMethodSite.java:48)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite$PogoCachedMethodSite.invoke(PogoMetaMethodSite.java:166)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite.call(PogoMetaMethodSite.java:69)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:148)
	at nextflow.processor.TaskRun.containerInfo0(TaskRun.groovy:626)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:362)
	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.callCurrent(PogoMetaClassSite.java:61)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:176)
	at nextflow.processor.TaskRun$_containerInfo_closure7.doCall(TaskRun.groovy:612)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:274)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1035)
	at groovy.lang.Closure.call(Closure.java:412)
	at org.codehaus.groovy.runtime.ConvertedClosure.invokeCustom(ConvertedClosure.java:50)
	at org.codehaus.groovy.runtime.ConversionHandler.invoke(ConversionHandler.java:112)
	at jdk.proxy1/jdk.proxy1.$Proxy39.apply(Unknown Source)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1708)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.codehaus.groovy.runtime.callsite.PlainObjectMetaMethodSite.doInvoke(PlainObjectMetaMethodSite.java:48)
	at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite$PojoCachedMethodSiteNoUnwrap.invoke(PojoMetaMethodSite.java:198)
	at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite.call(PojoMetaMethodSite.java:51)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:148)
	at nextflow.processor.TaskRun.containerInfo(TaskRun.groovy:612)
	at nextflow.processor.TaskRun.getContainer(TaskRun.groovy:634)
	at nextflow.processor.TaskRun.isContainerEnabled(TaskRun.groovy:679)
	at nextflow.processor.TaskRun$isContainerEnabled$10.call(Unknown Source)
	at nextflow.processor.TaskProcessor.createTaskHashKey(TaskProcessor.groovy:2070)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.codehaus.groovy.runtime.callsite.PlainObjectMetaMethodSite.doInvoke(PlainObjectMetaMethodSite.java:48)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite$PogoCachedMethodSiteNoUnwrapNoCoerce.invoke(PogoMetaMethodSite.java:189)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite.callCurrent(PogoMetaMethodSite.java:57)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:185)
	at nextflow.processor.TaskProcessor.invokeTask(TaskProcessor.groovy:617)
	at nextflow.processor.InvokeTaskAdapter.call(InvokeTaskAdapter.groovy:52)
	at groovyx.gpars.dataflow.operator.DataflowOperatorActor.startTask(DataflowOperatorActor.java:120)
	at groovyx.gpars.dataflow.operator.ForkingDataflowOperatorActor.access$001(ForkingDataflowOperatorActor.java:35)
	at groovyx.gpars.dataflow.operator.ForkingDataflowOperatorActor$1.run(ForkingDataflowOperatorActor.java:58)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
May-23 02:40:30.488 [Actor Thread 38] DEBUG nextflow.Session - Session aborted -- Cause: Wave does not support 'null' container engine
May-23 02:40:30.548 [Actor Thread 3] DEBUG nextflow.processor.TaskProcessor - Handling unexpected condition for
  task: name=NFCORE_SAREK:SAREK:PREPARE_INTERVALS:GATK4_INTERVALLISTTOBED (genome); work-dir=s3://seqeralabs-showcase/work/sarek/work-20230523/24/b4db8cc9bd0c3507114ab07c3a987f
  error [java.lang.IllegalArgumentException]: Wave does not support 'null' container engine
May-23 02:40:30.548 [Actor Thread 33] DEBUG nextflow.processor.TaskProcessor - Handling unexpected condition for
  task: name=NFCORE_SAREK:SAREK:PREPARE_INTERVALS:GATK4_INTERVALLISTTOBED; work-dir=null
  error [java.lang.InterruptedException]: java.lang.InterruptedException
  ```

### Environment 

* Nextflow version: 23.04.1 build 5866
* Java version: [?]
* Operating system: [macOS, Linux, etc]
* Bash version: (use the command `$SHELL --version`)

### Additional context

(Add any other context about the problem here)

**Comment by pditommaso** [comment]
It looks that for some reason it's unable to find the container engine here 

https://github.com/nextflow-io/nextflow/blob/9d9e275889712a4767f26c8fbb151ef3b7b912d2/plugins/nf-wave/src/main/io/seqera/wave/plugin/resolver/WaveContainerResolver.groovy#L69-L69

**Comment by pditommaso** [comment]
Tentative fix f227f2e5

**Comment by pditommaso** [comment]
This has also been observed in the latest 23.06.0-edge version including the above patch

**Comment by bentsherman** [comment]
In both cases, the error happens very quickly. It happens when the task processor tries to create the task hash before executing the task, so the first few tasks fail with this error.

On AWS Batch, this method will always be called with `docker`: https://github.com/nextflow-io/nextflow/blob/cb9592063783576a10ec78800e0ce99ff459bfb8/modules/nextflow/src/main/groovy/nextflow/Session.groovy#L1175-L1176

Is `Memoized` thread-safe? I assume it's using a ConcurrentHashMap under the hood. Of course if it wasn't, I would expect a different kind of error.

We might just have to add some `log.debug` messages to this code, wait for someone to encounter it again, and see where it goes wrong.

**Comment by pditommaso** [comment]
Don't think  @Memoized  is thread-safe

**Comment by pditommaso** [comment]
Solved by f3eba3d7
